lib/vdsm/storage/sdm/api/copy_data.py

```diff
@@ -146,15 +141,21 @@ class CopyDataDivEndpoint(properties.Owner):
                                     maxval=sc.MAX_GENERATION)
     prepared = properties.Boolean(default=False)
 
-    def __init__(self, params, host_id, writable, is_destination=False,
-                 lock_image=True):
+    def __init__(self, params, host_id, writable, dest=None):
         self.sd_id = params.get('sd_id')
         self.img_id = params.get('img_id')
         self.vol_id = params.get('vol_id')
         self.generation = params.get('generation')
         self.prepared = params.get('prepared')
-        self.is_destination = is_destination
-        self.lock_image = lock_image
+        self.is_destination = dest is None
+
+        # Always lock the destination image and only lock the source if
+        # the copy is not within the same image.
+        self.lock_image = (
+            self.is_destination
+            or getattr(dest, "img_id", None) != self.img_id
```

Comment:
Since you ensure that external endpoint always have an img_id, we can simplify this to:

    self.lock_image = self.is_destination or self.img_id != dest.img_id